### ISSUE TITLE
Adding variable to metadata

### ISSUE THREAD

**Issue by mstagliamonte** [joey711/phyloseq#1106]
Hi,

This is not really a big deal, more of a convenience issue. I was wondering if it possible to add a variable of choice to the metadata; as an example, if I forgot to add it to the mapping file, can I add the extra variable to the phyloseq object, or do I need to re-import an edited mapping table?

Thank you for your kind attention,
Max

**Comment by mikemc** [comment]
Yes, this is possible. The sample_data object is a special container around a data frame, and you can add columns to the data frame just like any other. You need to make sure when doing so that the samples are in the same order as in the phyloseq object. For example, to add a column that records the total read depth of the sample you could do
```r
sample_data(ps)$Read_depth <- sample_sums(ps)
```
and this is ok because phyloseq makes sure that the samples are ordered in the same way in `sample_data(ps)` and `sample_sums(ps)`.

**Comment by mstagliamonte** [comment]
Hi, @mikemc , 

That's great. I am away from my work computer now, I will try your approach in a few days. I have the new metadata  just in a table. I could feed the new column just as a vector, I will just have to make sure the order is consistent and do a sanity check afterwards.  

Thank you for your help,
Max

**Comment by mikemc** [comment]
Hi Max, when you have your new metadata in a table with the sample names, but not necessarily in the same order, then a simple and safer way is to use the `merge_phyloseq()` function. Create a new `sample_data` object of your new sample data, and then merge it with the current `phyloseq` object. Here's an example with some demo data:
```r
library(phyloseq)
data(GlobalPatterns)
ps <- GlobalPatterns
# Fake new sample variable
set.seed(0)
sam.new <- data.frame(New_var = sample(LETTERS, size = nsamples(ps), replace = TRUE))
# Mix up the sample names (just for demonstration purposes)
rownames(sam.new) <- sample(sample_names(ps))
head(sam.new)
#>          New_var
#> LMEpi24M       X
#> CL3            G
#> M11Tong        J
#> TRRsed3        O
#> F21Plmr        X
#> Even3          F
# Merge with original phyloseq object
ps0 <- merge_phyloseq(ps, sam.new)
head(sample_data(ps0))
#>         X.SampleID  Primer Final_Barcode Barcode_truncated_plus_T
#> CL3            CL3 ILBC_01        AACGCA                   TGCGTT
#> CC1            CC1 ILBC_02        AACTCG                   CGAGTT
#> SV1            SV1 ILBC_03        AACTGT                   ACAGTT
#> M31Fcsw    M31Fcsw ILBC_04        AAGAGA                   TCTCTT
#> M11Fcsw    M11Fcsw ILBC_05        AAGCTG                   CAGCTT
#> M31Plmr    M31Plmr ILBC_07        AATCGT                   ACGATT
#>         Barcode_full_length SampleType
#> CL3             CTAGCGTGCGT       Soil
#> CC1             CATCGACGAGT       Soil
#> SV1             GTACGCACAGT       Soil
#> M31Fcsw         TCGACATCTCT      Feces
#> M11Fcsw         CGACTGCAGCT      Feces
#> M31Plmr         CGAGTCACGAT       Skin
#>                                        Description New_var
#> CL3       Calhoun South Carolina Pine soil, pH 4.9       G
#> CC1       Cedar Creek Minnesota, grassland, pH 6.1       E
#> SV1     Sevilleta new Mexico, desert scrub, pH 8.3       Q
#> M31Fcsw    M3, Day 1, fecal swab, whole body study       R
#> M11Fcsw   M1, Day 1, fecal swab, whole body study        J
#> M31Plmr    M3, Day 1, right palm, whole body study       G
```

**Comment by mstagliamonte** [comment]
Hi, @mikemc ,

Thanks again for your detailed answer. I just tried and it works great! It's a very convenient set of functions you described, which I am probably going to use pretty often.

One thing I would highlight for the audience is that the row names of the data frame with the additional variables  need to  be named as per sample ID; that seems to be what the `merge_phyloseq` function uses as reference to merge. 

Thanks again, yours was a great answer.

Best,
Max

**Comment by mikemc** [comment]
One technical correction to what you say above @mstagliamonte: The row names of the data.frame for the new metadata need to match the row names of `sample_data(ps)` where `ps` is the existing phyloseq object---these are also what is returned by `sample_names(ps)`. It so happens that the `GlobalPatterns` phyloseq object also has this same information in a column of the `sample_data` called `X.sampleID` but this is not used by phyloseq during the merge and could be deleted without affecting anything.

**Comment by mstagliamonte** [comment]
Thank you for clarifying.  So, for what I understand,  the essential part is that row names between the phyloseq object and new metadata are to be consistent. I'll keep that in mind.

Best,
Max

**Comment by Thermoraptor** [comment]
If i have a phyloseq object and i identify certain taxa i want to subset.  I subset this taxa into a new phyloseq object. Is it possiblle to say label this subset degrading and then label the non-degrading taxa. Then compare ?

**Comment by augersam** [comment]
> Thank you for clarifying. So, for what I understand, the essential part is that row names between the phyloseq object and new metadata are to be consistent. I'll keep that in mind.
> 
> Best,
> Max

And, to further clarify, that "turn into sample data" step is critical otherwise it's not the proper object type and the merge_phyloseq will fail.
